File: vespalib/src/vespa/vespalib/data/slime/object_value.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "object_value.h"
#include "object_traverser.h"
#include "empty_value_factory.h"
#include "resolved_symbol.h"
#include "named_symbol_lookup.h"
#include "named_symbol_inserter.h"
#include "symbol_table.h"

namespace vespalib::slime {

Cursor &
ObjectValue::setLeaf(Symbol sym, const ValueFactory &input) {
    ResolvedSymbol symbol(sym);
    return setIfUnset(symbol, input);

}

Cursor &
ObjectValue::setLeaf(Memory name, const ValueFactory &input) {
    NamedSymbolInserter symbol(_symbolTable, name);
    return setIfUnset(symbol, input);
}

void
ObjectValue::traverse(ObjectSymbolTraverser &ot) const {
    for (const auto & field : _fields) {
        ot.field(field.first, *field.second);
    }
}

void
ObjectValue::traverse(ObjectTraverser &ot) const {
    for (const auto & field : _fields) {
        Memory symbol = _symbolTable.inspect(field.first);
        ot.field(symbol, *field.second);
    }
}

Cursor &
ObjectValue::operator[](Symbol sym) const {
    ResolvedSymbol symbol(sym);
    return *lookup(symbol);
}

Cursor &
ObjectValue::operator[](Memory name) const {
    NamedSymbolLookup symbol(_symbolTable, name);
    return *lookup(symbol);
}


Cursor &
ObjectValue::setArray(Symbol symbol, size_t reserve) {
    return setLeaf(symbol, ArrayValueFactory(_symbolTable, reserve));
}

Cursor &
ObjectValue::setObject(Symbol symbol) {
    return setLeaf(symbol, ObjectValueFactory(_symbolTable));
}

Cursor &
ObjectValue::setArray(Memory name, size_t reserve) {
    return setLeaf(name, ArrayValueFactory(_symbolTable, reserve));
}

Cursor &
ObjectValue::setObject(Memory name) {
    return setLeaf(name, ObjectValueFactory(_symbolTable));
}

Symbol
ObjectValue::resolve(Memory symbol_name) { return _symbolTable.insert(symbol_name); }

} // namespace vespalib::slime

File: vespajlib/src/main/java/com/yahoo/slime/ObjectValue.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.slime;

/**
 * A Value holding a slime "Object", a dynamic collection of named
 * value fields.  Fields can be inspected or traversed using the
 * {@link Inspector} interface, and you can add new fields by using the
 * various "set" methods in the @ref Cursor interface.
 *
 * @author havardpe
 */
final class ObjectValue extends Value {

    private int capacity = 16;
    private int hashSize() { return (capacity + (capacity >> 1) - 1); }
    private int used = 0;
    private Value[] values = new Value[capacity];
    private int[] hash = new int[capacity + hashSize() + (capacity << 1)];
    private final SymbolTable names;

    private void rehash() {
        capacity = (capacity << 1);
        Value[] v = values;
        values = new Value[capacity];
        System.arraycopy(v, 0, values, 0, used);
        int[] h = hash;
        hash = new int[capacity + hashSize() + (capacity << 1)];
        System.arraycopy(h, 0, hash, 0, used);
        for (int i = 0; i < used; i++) {
            int prev = (capacity + (hash[i] % hashSize()));
            int entry = hash[prev];
            while (entry != 0) {
                prev = entry + 1;
                entry = hash[prev];
            }
            final int insertIdx = (capacity + hashSize() + (i << 1));
            hash[prev] = insertIdx;
            hash[insertIdx] = i;
        }
    }

    private Value put(int sym, Value value) {
        if (used == capacity) {
            rehash();
        }
        int prev = (capacity + (sym % hashSize()));
        int entry = hash[prev];
        while (entry != 0) {
            final int idx = hash[entry];
            if (hash[idx] == sym) { // found entry
                return NixValue.invalid();
            }
            prev = entry + 1;
            entry = hash[prev];
        }
        final int insertIdx = (capacity + hashSize() + (used << 1));
        hash[prev] = insertIdx;
        hash[insertIdx] = used;
        hash[used] = sym;
        values[used++] = value;
        return value;
    }

    private Value get(int sym) {
        int entry = hash[capacity + (sym % hashSize())];
        while (entry != 0) {
            final int idx = hash[entry];
            if (hash[idx] == sym) { // found entry
                return values[idx];
            }
            entry = hash[entry + 1];
        }
        return NixValue.invalid();
    }

    public ObjectValue(SymbolTable names) { this.names = names; }
    public ObjectValue(SymbolTable names, int sym, Value value) {
        this.names = names;
        put(sym, value);
    }

    public Type type() { return Type.OBJECT; }
    public int children() { return used; }
    public int fields() { return used; }

    public Value field(int sym) { return get(sym); }
    public Value field(String name) { return get(names.lookup(name)); }

    public void accept(Visitor v) { v.visitObject(this); }

    public void traverse(ObjectSymbolTraverser ot) {
        for (int i = 0; i < used; ++i) {
            ot.field(hash[i], values[i]);
        }
    }

    public void traverse(ObjectTraverser ot) {
        for (int i = 0; i < used; ++i) {
            ot.field(names.inspect(hash[i]), values[i]);
        }
    }

    protected Cursor setLeaf(int sym, Value value) { return put(sym, value); }
    public Cursor setArray(int sym) { return put(sym, new ArrayValue(names)); }
    public  Cursor setObject(int sym) { return put(sym, new ObjectValue(names)); }

    protected Cursor setLeaf(String name, Value value) { return put(names.insert(name), value); }
    public Cursor setArray(String name) { return put(names.insert(name), new ArrayValue(names)); }
    public Cursor setObject(String name) { return put(names.insert(name), new ObjectValue(names)); }

}
